nixos/modules/programs/direnv.nix

```diff
@@ -0,0 +1,106 @@
+{
+  lib,
+  config,
+  pkgs,
+  ...
+}: let
+  cfg = config.programs.direnv;
+in {
+  options.programs.direnv = {
+
+    enable = lib.mkEnableOption (lib.mdDoc ''
+      direnv integration. Takes care of both installation and
+      setting up the sourcing of the shell. Additionally enables nix-direnv
+      integration. Note that you need to logout and login for this change to apply.
+    '');
+
+    package = lib.mkPackageOptionMD pkgs "direnv" {};
+
+    direnvrcExtra = lib.mkOption {
+      type = lib.types.lines;
+      default = "";
+      example = ''
+        export FOO="foo"
+        echo "loaded direnv!"
+      '';
+      description = lib.mdDoc ''
+        Extra lines to append to the sourced direnvrc
+      '';
+    };
+
+    silent = lib.mkEnableOption (lib.mdDoc ''
+      the hiding of direnv logging
+    '');
+
+    persistDerivations =
+      (lib.mkEnableOption (lib.mdDoc ''
+        setting keep-derivations and keep-outputs to true
+        to prevent shells from getting garbage collected
+      ''))
+      // {
+        default = true;
+      };
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    programs = {
+      zsh.interactiveShellInit = ''
+        eval "$(${lib.getExe cfg.package} hook zsh)"
+      '';
+
+      bash.interactiveShellInit = ''
+        eval "$(${lib.getExe cfg.package} hook bash)"
+      '';
+
+      fish.interactiveShellInit = ''
+        ${lib.getExe cfg.package} hook fish | source;
+      '';
+    };
+
+    nix.settings = lib.mkIf cfg.persistDerivations {
+      keep-outputs = true;
+      keep-derivations = true;
+    };
+
+    environment = {
+
+      systemPackages = [cfg.package];
+
+      variables = {
+        DIRENV_CONFIG = "/etc/direnv";
+        DIRENV_LOG_FORMAT = lib.mkIf cfg.silent "";
+      };
+
+      etc = {
+        "direnv/direnvrc".text = ''
+          #Load nix-direnv
+          source ${pkgs.nix-direnv}/share/nix-direnv/direnvrc
```

Comment:
Maybe we should also let `nix-direnv` be configured in the options ? WDYT?

Something like:

```nix
package-nix-direnv = lib.mkPackageOptionMD pkgs "nix-direnv" {};
```

Reply:
Finding the right path to source wouldn't be easy... If anything I'd remove the sourcing of `nix-direnv` and say use `direnvrcExtraConfig`